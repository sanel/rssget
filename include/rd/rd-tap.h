#ifndef __RD_TAP_H__
#define __RD_TAP_H__

/* TAP - test anything protocol */

void rd_plan(int n);
int  rd_done(void);

void rd_todo(const char *msg);
void rd_todo_end();

void rd_skip(int n, const char *msg);
void rd_skip_end();

int _rd_ok(int expr, const char *msg, const char *path, unsigned int line);
int _rd_is_str(const char *s1, const char *s2, const char *msg, const char *path, unsigned int line);

#define rd_ok(expr, str)        _rd_ok(expr, str, __FILE__, __LINE__)
#define rd_not_ok(expr, str)    _rd_ok(!(expr), str, __FILE__, __LINE__)
#define rd_is_str(s1, s2, str)  _rd_is_str(s1, s2, str, __FILE__, __LINE__)

#endif

#if 0
rd_plan(5);

rd_ok(1 == 1, "check if 1 eq 1");
rd_not_ok(1 == 2, "check 1 not 2");

rd_skip(rd_ok(1 == 1, 0), 2);
rd_ok(1 + 2, 0);
rd_ok(1 + 2, 0);
rd_ok(1 + 2, 0);
rd_ok(1 + 2, 0);
rd_skip_end();

rd_done();
#endif
