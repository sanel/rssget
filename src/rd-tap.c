#include <rd/rd-tap.h>
#include <stdio.h>
#include <string.h>

static int texpected;
static int tfailed;
static int tcurr;
static const char *todo_msg;

void rd_plan(int n) {
	texpected = n;
	printf("1..%i\n", n);
}

int rd_done(void) {
	int ret = 0;

	if (texpected > tcurr)  {
		printf("# Looks like you have planned %i but executed %i tests\n", texpected, tcurr);
		ret = 1;
	}

	if (tfailed) {
		printf("# Looks like you have failed %i tests out of %i run and %i planned\n",
			   tfailed, tcurr, texpected);
		ret = 1;
	}

	return ret;
}

int _rd_ok(int expr, const char *name, const char *path, unsigned int line) {
	if (!expr) printf("not ");
	printf("ok %i", ++tcurr);
	if (name) printf(" - %s", name);
	if (todo_msg) printf(" # TODO %s", todo_msg);

	printf("\n");

	if (!expr) {
		printf("#   Failed");
		if (todo_msg) printf("(TODO) ");
		printf("test ");
		if (name) printf("'%s'\n#   ", name);
		printf("at %s line %i.\n", path, line);
		if(!todo_msg)
			tfailed++;
	}

	return expr;
}

int _rd_is_str(const char *s1, const char *s2, const char *msg, const char *path, unsigned int line) {
	return _rd_ok(strcmp(s1, s2) == 0, msg, path, line);
}

void rd_todo(const char *msg) {
	todo_msg = msg;
}

void rd_todo_end(void) {
	todo_msg = 0;
}
