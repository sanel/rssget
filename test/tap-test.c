#include <rd/rd-tap.h>

int main() {
	rd_plan(3);

	rd_ok(1, "Sample 1");
	rd_ok(1, "Sample 2");
	rd_not_ok(2 == 3, "should 2 != 3");

	rd_todo("Should be skipped");
	rd_ok(0, "Failed TODO");
	rd_ok(1, "Sample 3");
	rd_is_str("foo", "foo", "foo == foo");
	rd_todo_end();

	return rd_done();
}
