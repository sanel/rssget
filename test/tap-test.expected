1..3
ok 1 - Sample 1
ok 2 - Sample 2
ok 3 - should 2 != 3
not ok 4 - Failed TODO # TODO Should be skipped
#   Failed(TODO) test 'Failed TODO'
#   at test/tap-test.c line 11.
ok 5 - Sample 3 # TODO Should be skipped
ok 6 - foo == foo # TODO Should be skipped
